Hide function prototyes from unauthorized callers

khwilliamson opened the pull request:
0351a629e71de127cbfd1b142e9eaa6069deabf5 extended hiding private  functions from callers into the gcc world.
     
Some functions are allowed only in extensions; so can not be marked as hidden; this commit discourages their use however, by hiding their prototypes to all but the core and extensions.
     
It turns out that four functions were being used in modules we ship with that were marked as extensions-only; so they had to be made  globally accessible.

* This set of changes does not require a perldelta entry.

khwilliamson committed:
regen/embed.pl: Add comments

khwilliamson committed:
regen/embed.pl: Squash duplicate code

Move these to a place common to both

khwilliamson committed:
regen/embed.pl: A,C and either [EX] flags are mutually exclusive

It makes no sense to specify more than one of A C X, nor more than one
of A C E.  This commit enforces that, showing one entry in embed.fnc
that needed to be changed.

khwilliamson committed:
embed.fnc: Add some comments

khwilliamson committed:
Hide function prototyes from unauthorized callers

0351a629e71de127cbfd1b142e9eaa6069deabf5 extended hiding private
functions from callers into the gcc world.

Some functions are allowed only in extensions; so can not be marked as
hidden; this commit discourages their use however, by hiding their
prototypes to all but the core and extensions.

It turns out that four functions were being used in modules we ship
with that were marked as extensions-only; so they had to be made
globally accessible.

tonycoz commented on regen/embed.pl:
```diff
@@ -142,6 +142,14 @@ sub generate_proto_h {
             warn "It is nonsensical to require the return value of a void function ($plain_func) to be checked";
         }
 
+        my $count = 0;
+        $count++ while $flags =~ /[ACX]/g;
+        die_at_end "$plain_func: A, C, and X flags are mutually exclusive"
+                                                                if $count > 1;
+        $count = 0;
+        $count++ while $flags =~ /[ACE]/g;
+        die_at_end "$plain_func: A, C, and E flags are mutually exclusive"
+                                                                if $count > 1;
```
These could be simplified:
```
die_at_end "..."
    if $flags` =~ tr/ACX// > 1;
```
I thought the code was duplicated at first.

khwilliamson replied:
There were several pre-existing  places in the code where this could be used. I've added a commit to change those.

tonycoz commented:
This warning from building pretty much every XS with `-DPERL_RC_STACK`
```
../../inline.h: In function ‘Perl_rpp_invoke_xs’:
../../inline.h:1181:9: warning: implicit declaration of function ‘Perl_xs_wrap’; did you mean ‘Perl_runops_wrap’? [-Wimplicit-function-declaration]
 1181 |         Perl_xs_wrap(aTHX_ CvXSUB(cv), cv);
      |         ^~~~~~~~~~~~
      |         Perl_runops_wrap
```
and the cpphdrcheck failing for pretty much the same cause:
```
# cmd: c++ -c -DPERL_RC_STACK -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/home/runner/work/perl5/perl5/t/.. -I/home/runner/work/perl5/perl5/t/../lib/CORE source.cpp 2>&1
# out: In file included from /home/runner/work/perl5/perl5/t/../perl.h:7956,
#                  from source.cpp:4:
# /home/runner/work/perl5/perl5/t/../inline.h: In function ‘void Perl_rpp_invoke_xs(CV*)’:
# /home/runner/work/perl5/perl5/t/../inline.h:1181:9: error: ‘Perl_xs_wrap’ was not declared in this scope
#  1181 |         Perl_xs_wrap(aTHX_ CvXSUB(cv), cv);
#       |         ^~~~~~~~~~~~

```

bulk88 commented on embed.fnc:
```diff
@@ -2063,7 +2069,7 @@ Cp	|const char *|moreswitches					\
 				|NN SV *coderef 			\
 				|NULLOK SV *args
-CRTXip	|char * |mortal_getenv	|NN const char *str
+CRTip	|char * |mortal_getenv	|NN const char *str
```
All of libperl's `*getenv*()` are messy, especially WinPerl's sandwich of functions and macros. 3 identical SV mortals/Newx blocks are made each time on WinPerl for no reason right now since Perl mid-5.30s.

I think but im not sure unless i do a deep dive study, that this function is one of the functions that will become semi-public CPAN API if WinPerl becomes MS UTF16-aware in the near future.

tonycoz replied:
That isn't relevant to this change.

bulk88 commented on embed.fnc:
```diff
@@ -3681,9 +3687,9 @@ Cp	|UV	|_to_utf8_upper_flags					\
+Xop	|bool	|try_amagic_bin |int method				\
 				|int flags
-EXop	|bool	|try_amagic_un	|int method				\
+Xop	|bool	|try_amagic_un	|int method				\
```
`B::C` and `B::CC` style modules may might want these symbols, but I have wanted to not-delete them and do major plastic surgery on their C prototypes/parameter order/parameter meanings for many months now.

The branch is 40% done and not syntax error-ing on 1 on these `overload.pm` getter callouts, I'm targeting `-X` file test operators first in my branch, because they have the worst abstraction split, between what the pp_*() func is responsible for and what the Public-to-CORE-only front-end C functions of `overload.pm` API are responsible for.

its mostly about this string/expression `PL_op->op_type` and is it caller function (the `pp_*()s`) or callee functions (overload.pm front ends) that derefs `my_perl->Iop->op_type` and do a tree of if/else branch logic on the U16 inside `op_type`

bulk88 replied:
> `B::C` and `B::CC` style modules may might want these symbols, but I have wanted to not-delete them and do major plastic surgery on their C prototypes/parameter order/parameter meanings for many months now.

Would there is any interest from Unix only P5P ppl, to make the Win32/Win64 GH CI runners, dump the real symbol table of `perl5xx.dll` to STDERR/STDOUT  inside a Win32/Win64 GH CI runner on each build?

Its 1279 lines (C symbols) long on `perl543.dll` right now.

tonycoz replied:
> `B::C` and `B::CC` style modules may might want these symbols, but I have wanted to not-delete them and do major plastic surgery on their C prototypes/parameter order/parameter meanings for many months now.

The undocumented (but useful for ops) macros that call these use the long names, so I think they're still usable from tools like B::CC.

> Would there is any interest from Unix only P5P ppl, to make the Win32/Win64 GH CI runners, dump the real symbol table of perl5xx.dll to STDERR/STDOUT inside a Win32/Win64 GH CI runner on each build?

Generating an artifact would probably be better if we did capture this.

Though I'm not sure it would be useful without something to track it over time.

khwilliamson committed:
regen/embed.pl: Use tr///, /[]/ to save operations

I, and previous authors had forgotten that tr returns a count of
matching characters, so can tell you if more than one character in the
input string matches; as well as if none do.

This also uses a character class to save an operation.

Spotted by Tony Cook

tonycoz commented on regen/embed.pl:
```diff
@@ -135,7 +135,7 @@ sub generate_proto_h {
         my $has_mflag = ( $flags =~ /m/ );
         my $is_malloc = ( $flags =~ /a/ );
-        my $can_ignore = ( $flags !~ /R/ ) && ( $flags !~ /P/ ) && !$is_malloc;
+        my $can_ignore = ($flags =~ tr/RP// == 0) && !$is_malloc;
```
This could just be:
```
my $can_ignore = $flags !~ tr/RP// && !$is_malloc;
```
or perhaps better:
```
my $can_ignore = $flags !~ /[RP]/ && !$is_malloc;
```

khwilliamson replied:
I chose your "better" option
